pkg/aws/aws.go

```diff
@@ -124,65 +124,98 @@ type AwsProvider struct {
 	WorkingDirectory string
 }
 
-func GetSubnetID(ctx context.Context, provider *AwsProvider) (string, error) {
-	svc := ec2.NewFromConfig(provider.AwsConfig)
-
-	// first search for a default devpod specific subnet, if it fails
-	// we search the subnet with most free IPs that can do also public-ipv4
-	input := &ec2.DescribeSubnetsInput{
-		Filters: []types.Filter{
-			{
-				Name: aws.String("tag:devpod"),
-				Values: []string{
-					"devpod",
-				},
-			},
-		},
+func GetSubnet(ctx context.Context, provider *AwsProvider) (string, error) {
+	// in case a single subnet ID is specified, use it without further checks
+	if len(provider.Config.SubnetIDs) == 1 {
+		return provider.Config.SubnetIDs[0], nil
 	}
 
-	result, err := svc.DescribeSubnets(ctx, input)
-	if err != nil {
-		return "", err
-	}
+	svc := ec2.NewFromConfig(provider.AwsConfig)
+	// in case multiple subnet IDs are specified, we return the one with most free IPs
+	if len(provider.Config.SubnetIDs) > 1 {
+		subnets, err := svc.DescribeSubnets(ctx, &ec2.DescribeSubnetsInput{
+			SubnetIds: provider.Config.SubnetIDs,
+		})
+		if err != nil {
+			return "", fmt.Errorf("list specified subnets %q: %w", provider.Config.SubnetIDs, err)
+		}
+		var maxIPCount int32
+		var subnet *types.Subnet
+		for _, s := range subnets.Subnets {
+			if provider.Config.AvailabilityZone != "" && *s.AvailabilityZone != provider.Config.AvailabilityZone {
+				continue
+			}
+			if *s.AvailableIpAddressCount > maxIPCount {
+				maxIPCount = *s.AvailableIpAddressCount
+				subnet = &s
+			}
+		}
+
+		if subnet == nil {
+			if provider.Config.AvailabilityZone == "" {
+				return "", fmt.Errorf("no subnets found with IDs %q", provider.Config.SubnetIDs)
+			} else {
+				return "", fmt.Errorf("no subnets found with IDs %q in availability zone %q", provider.Config.SubnetIDs, provider.Config.AvailabilityZone)
+			}
+		}
 
-	if len(result.Subnets) > 0 {
-		return *result.Subnets[0].SubnetId, nil
+		return *subnet.SubnetId, nil
 	}
 
-	input = &ec2.DescribeSubnetsInput{
-		Filters: []types.Filter{
-			{
-				Name: aws.String("vpc-id"),
-				Values: []string{
-					provider.Config.VpcID,
-				},
-			},
+	// retrieve and index all visible subnets
+	input := &ec2.DescribeSubnetsInput{}
+	if provider.Config.AvailabilityZone != "" {
+		input.Filters = []types.Filter{
 			{
-				Name: aws.String("map-public-ip-on-launch"),
+				Name: aws.String("availability-zone"),
 				Values: []string{
-					"true",
+					provider.Config.AvailabilityZone,
 				},
 			},
-		},
+		}
 	}
+	p := ec2.NewDescribeSubnetsPaginator(svc, input)
```

Comment:
why do we paginate the response here but not above when describing subnets?

Reply:
Because in the first call, we're fetching the subnets by ID by adding a list of IDs to `DescribeSubnetsInput`. These subnet ids are user specified and the default page size is 1000. I doubt a user would provide more than 1000 subnet ids in the env variable.